crates/tlsn/src/prover.rs

```diff
@@ -202,95 +198,135 @@ impl Prover<state::CommitAccepted> {
         )
         .map_err(ProverError::config)?;
 
-        let (conn, conn_fut) = bind_client(socket, client);
+        let span = self.span.clone();
+        let ctrl = ProverControl {
+            mpc_ctrl: mpc_ctrl.clone(),
+        };
 
-        let fut = Box::pin({
-            let span = self.span.clone();
-            let mpc_ctrl = mpc_ctrl.clone();
-            async move {
-                let conn_fut = async {
-                    mux_fut
-                        .poll_with(conn_fut.map_err(ProverError::from))
-                        .await?;
+        let mpc_tls = MpcTlsClient::new(
+            mux_fut,
+            Box::new(mpc_fut.map_err(ProverError::from)),
+            keys,
+            vm,
+            span,
+            mpc_ctrl,
+            client,
+        );
 
-                    mpc_ctrl.stop().await?;
+        let prover = Prover::<state::Connected> {
+            config: self.config,
+            span: self.span,
+            state: state::Connected {
+                mux_ctrl,
+                ctrl,
+                server_name: config.server_name().clone(),
+                tls_client: Box::new(mpc_tls),
+            },
+        };
+        Ok(prover)
+    }
+}
 
-                    Ok::<_, ProverError>(())
-                };
+impl Prover<state::Connected> {
+    /// Returns `true` if the prover can read TLS data from the server.
+    pub fn wants_read_tls(&self) -> bool {
+        self.state.tls_client.wants_read_tls()
+    }
 
-                info!("starting MPC-TLS");
+    /// Returns `true` if the prover wants to write TLS data to the server.
+    pub fn wants_write_tls(&self) -> bool {
+        self.state.tls_client.wants_write_tls()
+    }
 
-                let (_, (mut ctx, tls_transcript)) = futures::try_join!(
-                    conn_fut,
-                    mpc_fut.in_current_span().map_err(ProverError::from)
-                )?;
+    /// Reads TLS data from the server.
+    pub fn read_tls(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
```

Comment:
Upon further thought, instead of using `std::io::Error` it probably makes more sense to use `ProverError`